### PR TITLE
EGamma calib/regression switch

### PR DESCRIPTION
**Pull request by dabelknap**
I added the ability to switch on and off the egamma calibrations and regressions.

### CODE REVIEW

**Comment by dabelknap** [comment]
This works fine when you are running with the corrections switched off, however I am having difficulty getting it to work with it turned on.

My original idea was to try to pass the `eleReg` flag into `patElectronProduction_cff.py` to remove the correction stuff from the sequence, but it didn't work. So, I am trying to add the electron correction modules to the sequence from within `patTupleProduction.py`, but this is not working either. I made an inline comment where the trouble is.

File: PatTools/python/patTupleProduction.py [hunk 1]
```diff
@@ -332,13 +332,41 @@ def configurePatTuple(process, isMC=True, **kwargs):
...
+         calibratedPatElectrons
+
     #setup the energy regression for the specific dataset
-    process.eleRegressionEnergy.energyRegressionType    = cms.uint32(2)
-    process.calibratedPatElectrons.correctionsType      = cms.int32(2)
-    process.calibratedPatElectrons.combinationType      = cms.int32(3)
-    process.calibratedPatElectrons.lumiRatio            = cms.double(1.0)
-    process.calibratedPatElectrons.synchronization      = cms.bool(True)
+    if kwargs['eleReg']:
+        final_electron_collection += eleRegressionEnergy
```
Comment (dabelknap) [inline PatTools/python/patTupleProduction.py]: This throws the error

```
TypeError: unsupported operand type(s) for +=: 'InputTag' and 'EDProducer'
```

**Comment by ekfriis** [comment]
Hi @dabelknap, did you ever get this working?

**Comment by dabelknap** [comment]
No, I haven't figured out the TypeError issue yet.

File: PatTools/python/patTupleProduction.py (hunk 1 above)
Comment (ekfriis) [inline PatTools/python/patTupleProduction.py]: I think the problem is that `final_electron_collection` is the item at the tail of customizeElectronSequence, not the sequence it self.  I think what you want to do is use the if to "insert" them into the cms.Sequence.  i.e. add 

``` python
if kwargs['eleReg']:
   process.customizeElectronSequence.insert(0, process.calibratedPatElectrons)
   process.customizeElectronSequence.insert(0, process.eleRegressionEnergy)
   # NB the order, as we are appending to front
```

This has to be before we paste the input electron collection, `selectedPatElectrons` into the start of the chain, @
https://github.com/dabelknap/FinalStateAnalysis/blob/397fed449e49047f3bae9530c98f04f391ea0e34/PatTools/python/patTupleProduction.py#L330

Does it make sense?

**Comment by ekfriis** [comment]
Hi Austin, any update on this?

**Comment by dabelknap** [comment]
I am still working on it. The changes I introduced created an unexpected error for the jet collections. I need to tinker with it some more.

**Comment by dabelknap** [comment]
The TypeError is no longer and issue, and the the jet collection seem unaffected, but I am getting a new error in `python -i patTuple_cfg.py eleReg=1`:

```
Traceback (most recent call last):
  File "patTuple_cfg.py", line 95, in <module>
    HLTprocess=options.HLTprocess, eleReg=options.eleReg
  File "/afs/hep.wisc.edu/cms/belknap/fsa539/python/FinalStateAnalysis/PatTools/patTupleProduction.py", line 346, in configurePatTuple
    process.customizeElectronSequence += cms.Sequence(eleRegressionEnergy)
  File "/cvmfs/cms.cern.ch/slc5_amd64_gcc462/cms/cmssw/CMSSW_5_3_9/python/FWCore/ParameterSet/Config.py", line 310, in __setattr__
    newValue._place(name,self)
  File "/cvmfs/cms.cern.ch/slc5_amd64_gcc462/cms/cmssw/CMSSW_5_3_9/python/FWCore/ParameterSet/SequenceTypes.py", line 130, in _place
    self._placeImpl(name,proc)
  File "/cvmfs/cms.cern.ch/slc5_amd64_gcc462/cms/cmssw/CMSSW_5_3_9/python/FWCore/ParameterSet/SequenceTypes.py", line 352, in _placeImpl
    proc._placeSequence(name,self)
  File "/cvmfs/cms.cern.ch/slc5_amd64_gcc462/cms/cmssw/CMSSW_5_3_9/python/FWCore/ParameterSet/Config.py", line 399, in _placeSequence
    self._validateSequence(mod, name)
  File "/cvmfs/cms.cern.ch/slc5_amd64_gcc462/cms/cmssw/CMSSW_5_3_9/python/FWCore/ParameterSet/Config.py", line 575, in _validateSequence
    raise RuntimeError("An entry in sequence "+label + ' has no label')
RuntimeError: An entry in sequence customizeElectronSequence has no label
```

I decided not to use `insert(0, ...)` since I wanted to run the electron corrections at the end of `customizeElectronSequence`.

**Comment by ekfriis** [comment]
Hi Austin,

This is a hard to debug error.  There is a module which has been associated to the cms.Process (giving it a "label", where the label is process.LABEL), but it has somehow been moved or deleted or given another label.  You should should look for some module which is "copied" without using either

``` python
import copy

my_new_copy = copy.deepcopy(the_module)
```

or 

``` python
my_new_copy = the_module.clone()
```

Let me know how it goes.

Evan

**Comment by dabelknap** [comment]
I think this is ready to merge. Let me know if you want me to clean up the history or anything.

**Pull request merged**